graphs/uniform_search_cost.py

```diff
@@ -0,0 +1,128 @@
+
+
+def uniform_search_cost(
+    graph_data: list[list[int]],
+    distance: dict[
+        tuple[
+            int,
+            int,
+        ],
+        int,
+    ],
+    goal: list[int],
+    start: int,
+) -> list[int]:
+    """
+    Returns a list of integer values which are result of optimal distance
+    >>> graph_data = [[1, 2], [2, 3], [], []]
+    >>> distance = {(0, 1): 2, (0, 2): 1, (1, 2): 5, (1, 3): 10}
+    >>> goal = [2]
+    >>> start = 0
+    >>> uniform_search_cost(graph_data, distance, goal, start)
+    [1]
+    """
+    # priority queue
+    heap = []
+    goal_answer = []
+    # Goal result to be max value initially
+    for _ in range(len(goal)):
+        goal_answer.append(10**8)
+
+    # Store the starting node
+    heap.append([0, start])
+
+    # store tracks of visited nodes
+    visited = {}
+    final_cnt = 0
+    while len(heap) > 0:
+        heap = sorted(heap)
+        # get the top element of the node list which has less distance value
+        current_node = heap[-1]
+        del heap[-1]
+        current_node[0] *= -1
+
+        # check current node is part of the goal list
+        if current_node[1] in goal:
+            index = goal.index(current_node[1])
+            # new goal is reached or not
+            if goal_answer[index] == 10**8:
+                final_cnt += 1
+            # if the distance is less
+            if goal_answer[index] > current_node[0]:
+                goal_answer[index] = current_node[0]
+            # pop the element
+            del heap[-1]
+            heap = sorted(heap)
+            # if it crosses all goals
+            if final_cnt == len(goal):
+                return goal_answer
+        # explore new unvisited node
+        if current_node[1] not in visited:
+            for i in range(len(graph_data[current_node[1]])):
+                # The value is negated to ensure that
+                # the node with the lowest priority is placed at the top.
+                heap.append(
+                    [
+                        (
+                            current_node[0]
+                            + distance[
+                                (current_node[1], graph_data[current_node[1]][i])
+                            ]
+                        )
+                        * -1,
+                        graph_data[current_node[1]][i],
+                    ]
+                )
+        visited[current_node[1]] = 1
+
+    return goal_answer
+
+
+def run() -> None:
+    """
+    Return nothing and prints only the output
+    >>> run()
+    start: 0, goal: 6; Distance(Minimum/Lowest): 1
+    """
+    # create the graph_data
+    graph_data = [[] for _ in range(4)]  # type: list[list[int]]
+
+    # Node to node edges
+    graph_data[0].append(1)
+    graph_data[0].append(2)
+    graph_data[1].append(2)
+    graph_data[1].append(3)
+
+    # Distance for the edges one node to another node
+    distance = {(0, 1): 2, (0, 2): 1, (1, 2): 5, (1, 3): 10}
```

Comment:
1) Why is the data for `graph_data` added after the list is instantiated? Why can't we just do `graph_data = [[1, 2], [2, 3]]`?
2) What graph representation is `graph_data`? Is it meant to be an adjacency list? If so, why are there no vertex lists for `2` and `3`?
3) You really shouldn't store the adjacency data and edge distances separately if you can avoid it. Is there a good reason to store the distances separately?

Reply:
Thanks for your queries. 

1. `graph_data = [[1, 2], [2, 3]]` represents a graph where nodes 0, 1, and 2 are connected in a linear fashion. I made it like a familiar readable way of node to node connection for static test case.
2. `graph_data` is meant to be an adjacency list representation of a graph. Do i need to add some comments about graph in the code?
3. Actually, i was storing adjacency data and edge distances separately is common in graph algorithms because it provides flexibility to me handling.

Reply:
@tianyizheng02, can you look over the latest change about the `graph_data` with test cases?

Reply:
> 1. `graph_data = [[1, 2], [2, 3]]` represents a graph where nodes 0, 1, and 2 are connected in a linear fashion. I made it like a familiar readable way of node to node connection for static test case.
> 2. `graph_data` is meant to be an adjacency list representation of a graph. Do i need to add some comments about graph in the code?

I think you're misunderstanding what an adjacency list is. What I suggested _is_ the (partly) correct adjacency list representation for the graph that you're proposing. An adjacency list contains a list of neighbors for each vertex; it does _not_ simply contain a list of edges.

Your graph has vertex 0 connected to vertices 1 and 2 and vertex 1 connected to vertices 2 and 3. In `graph_data = [[1, 2], [2, 3]]`, index 0 (which represents vertex 0) contains 1 and 2, and index 1 (which represents vertex 1) contains 2 and 3.

(Yes, the correct adjacency list should actually be `graph_data = [[1, 2], [2, 3], [], []]` to account for vertices 2 and 3, which have no neighbors. However, my point still stands that you can easily instantiate the entire adjacency list all at once.)

Also, yes, you should absolutely specify in the code that the input graph should be represented as an adjacency list.

> 3. Actually, i was storing adjacency data and edge distances separately is common in graph algorithms because it provides flexibility to me handling.

I strongly disagree with this. Firstly, in my experience it's not common to separate adjacency data and edge distances in graph representations (though you can disagree with me on this since it's anecdotal). However, it's still definitely worse IMO to decouple the adjacency data from the edge distances. Needing to update two separate data structures whenever you ever modify the graph makes it more prone to mistakes. Graph adjacency list representations can absolutely store edge distances alongside the adjacency lists for each vertex.